node/lib/util/synthetic_gc_util.js

```diff
@@ -0,0 +1,330 @@
+    function* (repo, commit, isDeletable, existingReferences) {
+
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    if (commit.parentcount() === 0) {
+        return;
+    }
+
+    const parents = yield commit.getParents(commit.parentcount());
+    let thisState = this;
+    yield parents.map(function *(parent) {
+        if (parent.sha() in thisState.d_visited) {
+            return;
+        }
+        thisState.d_visited[parent.sha()] = 1;
+
+        // We are keeping track of 'existingReferences' to avoid trying to
+        // delete referneces that do not exists. This is helpful in simulation
+        // mode, since it provides a clear way to a user what actually is being
+        // deleted. Also helps in debugging.
+        if (isDeletable(parent) && existingReferences.includes(parent.sha())) {
+            thisState.removeSyntheticRef(repo, parent);
+        }
+        return yield thisState.recursiveSyntheticRefRemoval(repo, parent,
+                                                   isDeletable,
+                                                   existingReferences);
+    });
+
+});
+
+/**
+ * Return all available synthetic refs within specified `subRepo`.
+ *
+ * @param {NodeGit.Repo}   subRepo
+ * @return {String[]}
+ */
+SyntheticGcUtil.prototype.getSyntheticRefs= co.wrap(
+    function*(subRepo) {
+
+    assert.instanceOf(subRepo, NodeGit.Repository);
+    let references = yield NodeGit.Reference.list(subRepo);
+
+    references = references.filter(
+            ref => ref.indexOf(SYNTHETIC_BRANCH_BASE) !== -1);
+
+    references = references.map(ref => ref.split(SYNTHETIC_BRANCH_BASE)[1]);
+
+    return references;
+});
+
+/**
+ * Delete all redundant synthetic refs within specified 'repo' satisfying
+ * `predicate` by recursively iterating over parents of the specified `roots`.
+ *
+ * Synthetic ref is considered to be redundant if its commit is reachable from
+ * descendant who is guaranteed to be around - i.e part of a persistent roots
+ * ('roots' here).
+ *
+ * @param {NodeGit.Repo}   repo
+ * @param {Object[]}       roots
+ * @param {Function}       predicate
+ */
+SyntheticGcUtil.prototype.cleanUpRedundant= co.wrap(
+    function*(repo, roots, predicate) {
+
+   assert.instanceOf(repo, NodeGit.Repository);
+
+   for (let subPath in roots) {
+
+       const subRepo = yield NodeGit.Repository.open(subPath);
+
+       let existingReferences = yield this.getSyntheticRefs(subRepo);
+
+       for (let subCommit of roots[subPath]) {
+           yield this.recursiveSyntheticRefRemoval(subRepo, subCommit,
+                                              predicate,
+                                              existingReferences);
+       }
+   }
+});
+
+/**
+ * Delete all synthetic refs within specified `repo` that satisfy `isOldCommit`,
+ * that not part of specified `roots`.
+ *
+ * @param {NodeGit.Repo}   repo
+ * @param {Object[]}       roots
```

Comment:
I don't think roots is really an array -- I think it's a map from string to array, right?